envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -163,23 +210,57 @@ class RBACFilterFactory(
         }
     }
 
-    private fun sourceIpPrincipals(client: String, snapshot: GlobalSnapshot): List<Principal> {
+    private fun ipFromDiscoveryPrincipals(
+        client: Client,
+        selector: String?,
+        selectorMatching: SelectorMatching?,
+        snapshot: GlobalSnapshot
+    ): List<Principal> {
         val clientEndpoints = snapshot.endpoints.resources().filterKeys { client == it }.values
         return clientEndpoints.flatMap { clusterLoadAssignment ->
             clusterLoadAssignment.endpointsList.flatMap { lbEndpoints ->
                 lbEndpoints.lbEndpointsList.map { lbEndpoint ->
                     lbEndpoint.endpoint.address
                 }
             }
-        }.map { address ->
-            Principal.newBuilder().setSourceIp(CidrRange.newBuilder()
+        }.flatMap { address ->
+            val sourceIpPrincipal = Principal.newBuilder()
+                    .setSourceIp(CidrRange.newBuilder()
                     .setAddressPrefix(address.socketAddress.address)
                     .setPrefixLen(EXACT_IP_MASK).build())
                     .build()
+
+            if (selector != null && selectorMatching != null) {
+                addAdditionalMatching(selector, selectorMatching, listOf(sourceIpPrincipal))
```

Comment:
I think we should do it the same as for static ip ranges:
```
(sourceIp1 OR sourceIp2) AND (headerMatching)
```
instead of:
```
(sourceIp1 AND headerMatching) OR (sourceIp2 AND headerMatching)
```